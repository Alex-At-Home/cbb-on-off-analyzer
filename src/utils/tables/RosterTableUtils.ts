// Lodash
import _ from "lodash";

// Util imports
import { RatingUtils } from "../stats/RatingUtils";
import { PositionUtils } from "../stats/PositionUtils";
import { LineupUtils } from "../stats/LineupUtils";
import { LuckUtils, OffLuckAdjustmentDiags, DefLuckAdjustmentDiags, LuckAdjustmentBaseline } from "../stats/LuckUtils";

/** Object marshalling logic for roster tables */
export class RosterTableUtils {

  /** Build a lookup map of the roster by their code, eg AaWiggins for "Wiggins, Aaron" (teamSeasonLookup only needed if includePosCat is true) */
  static buildRosterTableByCode(players: Array<any>, rosterInfo: Record<string, any> | undefined, includePosCat?: boolean, teamSeasonLookup?: string) {
    return _.chain(players).map(p => {
      const code = p.player_array?.hits?.hits?.[0]?._source?.player?.code || p.key;
      const rosterEntry = rosterInfo?.[code];
      // Ugly but we mutate p here:
      if  (rosterEntry) p.roster = rosterEntry;

      //TODO: do I want this one, or the baseline or ???
      if (includePosCat) {
        const [ posConfs, posConfsDiags ] = PositionUtils.buildPositionConfidences(p, rosterEntry?.height_in);
        const [ pos, posDiags ] = PositionUtils.buildPosition(posConfs, posConfsDiags.confsNoHeight, p, teamSeasonLookup || "");
        p.posClass = _.values(posConfs);
        p.role = pos;
      }

      return [ code, p ];
    }).fromPairs().value();

  }
};
