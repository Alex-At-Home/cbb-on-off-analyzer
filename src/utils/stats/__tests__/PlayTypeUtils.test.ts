import _ from 'lodash';

import { PlayTypeUtils } from "../PlayTypeUtils";
import { RosterTableUtils } from "../../tables/RosterTableUtils";
import { GameFilterParams, LineupFilterParams, TeamReportFilterParams } from "../../FilterModels";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleLineupStatsResponse } from "../../../sample-data/sampleLineupStatsResponse";

describe("PlayTypeUtils", () => {

  const teamSeasonLookup = "Men_Maryland_2018/9";
  const players =
    samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets || [];
  const rosterStatsByCode = RosterTableUtils.buildRosterTableByCode(
    players, {}, true, teamSeasonLookup
  );
  const mainPlayer = players[0];
  const allPlayers = PlayTypeUtils.buildPlayerAssistCodeList(mainPlayer);

  test("PlayTypeUtils - buildPlayerStyle", () => {

    const playStyle = PlayTypeUtils.buildPlayerStyle(mainPlayer);
    //expect(playStyle).toEqual({});
    expect(playStyle).toMatchSnapshot();

    const playStyleWithTotals = PlayTypeUtils.buildPlayerStyle(mainPlayer, 1, 1);
    //expect(playStyleWithTotals).toEqual({});
    expect(playStyle).toMatchSnapshot();
  });
  test("PlayTypeUtils - buildPlayerAssistNetwork", () => {
    const playerStyle = PlayTypeUtils.buildPlayerStyle(mainPlayer);
    const testPlayerCode = allPlayers.filter(p => rosterStatsByCode.hasOwnProperty(p))[0];

    const network = PlayTypeUtils.buildPlayerAssistNetwork(
      testPlayerCode, mainPlayer, playerStyle.totalScoringPlaysMade, playerStyle.totalAssists,
      rosterStatsByCode
    );
    //expect(network).toEqual({});
    expect(network).toMatchSnapshot();
  });
  test("PlayTypeUtils - enrichUnassistedStats", () => {
    const playerStyle = PlayTypeUtils.buildPlayerStyle(mainPlayer);
    const extraUnassistedInfo = PlayTypeUtils.enrichUnassistedStats(playerStyle.unassisted, mainPlayer);
    //expect(extraUnassistedInfo).toEqual({});
    expect(extraUnassistedInfo).toMatchSnapshot();

    const extraUnassistedInfoPos = PlayTypeUtils.enrichUnassistedStats(playerStyle.unassisted, 0);
    //expect(extraUnassistedInfoPos).toEqual({});
    expect(extraUnassistedInfoPos).toMatchSnapshot();
  });
  test("PlayTypeUtils - buildPosCategoryAssistNetwork", () => {
    const playerStyle = PlayTypeUtils.buildPlayerStyle(mainPlayer);
    const playerAssistNetwork = allPlayers.map((p) => {
      const [ info, ignore ] = PlayTypeUtils.buildPlayerAssistNetwork(
        p, mainPlayer, playerStyle.totalScoringPlaysMade, playerStyle.totalAssists,
        rosterStatsByCode
      );
      return info;
    });
    const posCategoryAssistNetwork = PlayTypeUtils.buildPosCategoryAssistNetwork(
      playerAssistNetwork, rosterStatsByCode, mainPlayer
    );
    //expect(posCategoryAssistNetwork).toEqual({});
    expect(posCategoryAssistNetwork).toMatchSnapshot();

    const posCategoryAssistNetworkNoInfo = PlayTypeUtils.buildPosCategoryAssistNetwork(
      playerAssistNetwork, rosterStatsByCode, undefined
    );
    //expect(posCategoryAssistNetworkNoInfo).toEqual({});
    expect(posCategoryAssistNetworkNoInfo).toMatchSnapshot();

    const posCategoryAssistNetworkPos = PlayTypeUtils.buildPosCategoryAssistNetwork(
      playerAssistNetwork, rosterStatsByCode, 0
    );
    //expect(posCategoryAssistNetworkPos).toEqual({});
    expect(posCategoryAssistNetworkPos).toMatchSnapshot();
  });
});
