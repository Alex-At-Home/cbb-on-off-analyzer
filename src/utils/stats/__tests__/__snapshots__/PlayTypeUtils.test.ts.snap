// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayTypeUtils PlayTypeUtils - aggregateToTopLevelPlayStyles 1`] = `
Object {
  "Attack & Kick": 0.1960333867510726,
  "Backdoor Cut": 0.048367783383114045,
  "Dribble Jumper": 0.06342965187105977,
  "Mid-Range": 0.053543789432893334,
  "Misc": 0.06465407003508915,
  "Put-Back": 0.0647724089708131,
  "Rim Attack": 0.1941475749678136,
  "Transition": 0.2048516551949357,
}
`;

exports[`PlayTypeUtils PlayTypeUtils - aggregateToTopLevelPlayStyles 2`] = `
Object {
  "Attack & Kick": 0.07365653548301296,
  "Backdoor Cut": 0.026352671583922415,
  "Dribble Jumper": 0.02455217849433765,
  "Mid-Range": 0.011457683297357569,
  "Misc": 0,
  "Put-Back": 0.03069022311792207,
  "Rim Attack": 0.08495053759040828,
  "Transition": 0.10148233777659563,
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildCategorizedAssistNetworks 1`] = `
Object {
  "ballhandler": Object {
    "assists": Array [
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": Array [
            "drive-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.005462550917096372,
        },
        "source_mid_ast": Object {
          "extraInfo": Array [
            "spread offense",
            "misc action",
          ],
          "value": 0.0007534552989098443,
        },
        "source_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.001883638247274611,
        },
        "target_3p_ast": Object {
          "extraInfo": Array [
            "drive-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.014180929095354525,
        },
        "target_3p_efg": Object {
          "extraInfo": undefined,
          "value": 0.4781305403461805,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.008099644463280826,
        },
        "target_mid_ast": Object {
          "extraInfo": Array [
            "spread offense",
            "misc action",
          ],
          "value": 0.0019559902200489,
        },
        "target_mid_efg": Object {
          "extraInfo": undefined,
          "value": 0.2432432432432432,
        },
        "target_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.004889975550122249,
        },
        "target_rim_efg": Object {
          "extraInfo": undefined,
          "value": 0.6649717514124294,
        },
      },
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": Array [
            "slash-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.03673094582185492,
        },
        "source_mid_ast": Object {
          "extraInfo": Array [
            "spread offense",
            "misc action",
          ],
          "value": 0.0003767276494549222,
        },
        "source_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.017706199524381343,
        },
        "target_3p_ast": Object {
          "extraInfo": Array [
            "drive-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.1677261613691932,
        },
        "target_3p_efg": Object {
          "extraInfo": undefined,
          "value": 0.11103158115256369,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.13863577499941135,
        },
        "target_mid_ast": Object {
          "extraInfo": Array [
            "spread offense",
            "misc action",
            "zone buster",
          ],
          "value": 0.02444987775061125,
        },
        "target_mid_efg": Object {
          "extraInfo": undefined,
          "value": 0.07783783783783783,
        },
        "target_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.16772616136919316,
        },
        "target_rim_efg": Object {
          "extraInfo": undefined,
          "value": 0.05811138014527844,
        },
      },
      Object {
        "order": 2,
      },
    ],
    "other": Array [
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.02053165689529326,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.019213110122201033,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.049539685903322264,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.10020955475500928,
        },
      },
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.042193496738951286,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.0011301829483647665,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.019589837771655953,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.1467354194626922,
        },
      },
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.009983282710555437,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.0016952744225471497,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.018459654823291188,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.04017329471874926,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.03786112877021968,
        },
      },
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.006027642391278755,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.0011301829483647665,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.00527418709236891,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.005815733088460361,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.009606555061100515,
        },
      },
      Object {
        "order": 0,
      },
      Object {
        "order": 0,
      },
    ],
  },
  "big": Object {
    "assists": Array [
      Object {
        "order": 0,
      },
      Object {
        "order": 1,
      },
      Object {
        "order": 2,
      },
    ],
    "other": Array [
      Object {
        "order": 2,
      },
      Object {
        "order": 2,
      },
      Object {
        "order": 2,
      },
      Object {
        "order": 2,
      },
      Object {
        "order": 2,
      },
      Object {
        "order": 2,
      },
    ],
  },
  "wing": Object {
    "assists": Array [
      Object {
        "order": 0,
        "source_3p_ast": Object {
          "extraInfo": Array [
            "drive-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.015069105978196887,
        },
        "source_mid_ast": Object {
          "extraInfo": Array [
            "spread offense",
            "misc action",
            "zone buster",
          ],
          "value": 0.0030138211956393774,
        },
        "source_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.005274187092368911,
        },
        "target_3p_ast": Object {
          "extraInfo": Array [
            "slash-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.0176039119804401,
        },
        "target_3p_efg": Object {
          "extraInfo": undefined,
          "value": 0.48275862068965514,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.009041463586918132,
        },
        "target_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.0058679706601467,
        },
        "target_rim_efg": Object {
          "extraInfo": undefined,
          "value": 0.5555555555555556,
        },
      },
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": Array [
            "slash-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.021850203668385484,
        },
        "source_mid_ast": Object {
          "extraInfo": Array [
            "spread offense, misc action",
            "zone buster",
          ],
          "value": 0.0015069105978196887,
        },
        "source_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.009041463586918132,
        },
        "target_3p_ast": Object {
          "extraInfo": Array [
            "slash-and-kick",
            "hockey assist after defense collapses inside",
            "misc action",
          ],
          "value": 0.0254278728606357,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.024110569565115023,
        },
        "target_mid_ast": Object {
          "extraInfo": Array [
            "spread offense, misc action",
            "zone buster",
          ],
          "value": 0.0078239608801956,
        },
        "target_rim_ast": Object {
          "extraInfo": Array [
            "cut",
          ],
          "value": 0.029339853300733493,
        },
      },
      Object {
        "order": 2,
      },
    ],
    "other": Array [
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.0030138211956393774,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.009041463586918132,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.01732947187492642,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.021473476018930564,
        },
      },
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.036919309646582375,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.004520731793459066,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.014315650679287042,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.033152033152033154,
        },
      },
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.009794918885827977,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.0030138211956393774,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.009794918885827977,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.008589390407572225,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.008288008288008289,
        },
      },
      Object {
        "order": 1,
        "source_3p_ast": Object {
          "extraInfo": undefined,
          "value": 0.0015069105978196887,
        },
        "source_mid_ast": Object {
          "extraInfo": undefined,
          "value": 0.0007534552989098443,
        },
        "source_rim_ast": Object {
          "extraInfo": undefined,
          "value": 0.006027642391278755,
        },
        "source_sf": Object {
          "extraInfo": undefined,
          "value": 0.0017894563349108803,
        },
        "target_ast": Object {
          "extraInfo": undefined,
          "value": 0.0007534552989098443,
        },
      },
      Object {
        "order": 1,
      },
      Object {
        "order": 1,
      },
    ],
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerAssistNetwork 1`] = `
Array [
  Object {
    "source_3p_ast": Object {
      "value": 0.02327385570209465,
    },
    "source_3p_efg": Object {
      "value": 0.4760479041916168,
    },
    "source_rim_ast": Object {
      "value": 0.007757951900698216,
    },
    "source_rim_efg": Object {
      "value": 0.711864406779661,
    },
    "target_3p_ast": Object {
      "value": 0.1360544217687075,
    },
    "target_3p_efg": Object {
      "value": 0.4760479041916168,
    },
    "target_ast": Object {
      "value": 0.08016550297388156,
    },
    "target_mid_ast": Object {
      "value": 0.027210884353741496,
    },
    "target_mid_efg": Object {
      "value": 0.24324324324324326,
    },
    "target_rim_ast": Object {
      "value": 0.047619047619047616,
    },
    "target_rim_efg": Object {
      "value": 0.711864406779661,
    },
  },
  43,
]
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerAssistNetwork 2`] = `
Array [
  Object {
    "source_3p_ast": Object {
      "value": 0.06982156710628394,
    },
    "source_3p_efg": Object {
      "value": 0.4760479041916168,
    },
    "source_rim_ast": Object {
      "value": 0.015515903801396431,
    },
    "source_rim_efg": Object {
      "value": 0.711864406779661,
    },
    "target_3p_ast": Object {
      "value": 0.40816326530612246,
    },
    "target_3p_efg": Object {
      "value": 0.4760479041916168,
    },
    "target_ast": Object {
      "value": 0.08016550297388156,
    },
    "target_mid_ast": Object {
      "value": 0.05442176870748299,
    },
    "target_mid_efg": Object {
      "value": 0.24324324324324326,
    },
    "target_rim_ast": Object {
      "value": 0.09523809523809523,
    },
    "target_rim_efg": Object {
      "value": 0.711864406779661,
    },
  },
  43,
]
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerStyle 1`] = `
Object {
  "assisted": Object {
    "source_3p_ast": Object {
      "value": 0.09050943884147919,
    },
    "source_mid_ast": null,
    "source_rim_ast": Object {
      "value": 0.04396172743728989,
    },
    "target_ast": Object {
      "value": 0.38013964313421256,
    },
  },
  "assistedMissAdjustments": undefined,
  "scramble": Object {
    "source_3p_ast": Object {
      "value": 0.015515903801396431,
    },
    "source_mid_ast": Object {
      "value": 0.002585983966899405,
    },
    "source_rim_ast": Object {
      "value": 0.01034393586759762,
    },
    "source_sf": Object {
      "value": 0.01474010861132661,
    },
    "target_ast": Object {
      "value": 0.025859839668994054,
    },
  },
  "scrambleAssisted": Object {},
  "totalAssists": 147,
  "totalPlaysMade": 386.7,
  "transition": Object {
    "source_3p_ast": Object {
      "value": 0.02068787173519524,
    },
    "source_mid_ast": Object {
      "value": 0.002585983966899405,
    },
    "source_rim_ast": Object {
      "value": 0.04396172743728989,
    },
    "source_sf": Object {
      "value": 0.10440910266356349,
    },
    "target_ast": Object {
      "value": 0.0982673907421774,
    },
  },
  "transitionAssisted": Object {},
  "unassisted": Object {
    "source_3p_ast": Object {
      "value": 0.05430566330488751,
    },
    "source_mid_ast": Object {
      "value": 0.0465477114041893,
    },
    "source_rim_ast": Object {
      "value": 0.12412723041117145,
    },
    "source_sf": Object {
      "value": 0.2604085854667701,
    },
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerStyle 2`] = `
Object {
  "assisted": Object {
    "source_3p_ast": Object {
      "value": 0.05202913631633715,
    },
    "source_mid_ast": null,
    "source_rim_ast": Object {
      "value": 0.0252712947822209,
    },
    "target_ast": Object {
      "value": 0.218522372528616,
    },
  },
  "assistedMissAdjustments": Object {
    "3p": 52.59946595460614,
    "mid": 1.2222222222222285,
    "rim": 8.563343491510672,
  },
  "scramble": Object {
    "source_3p_ast": Object {
      "value": 0.01932510777463951,
    },
    "source_mid_ast": Object {
      "value": 0.005946187007581388,
    },
    "source_rim_ast": Object {
      "value": 0.007432733759476735,
    },
    "source_sf": Object {
      "value": 0.008473316485803477,
    },
    "source_to": Object {
      "value": 0.002973093503790694,
    },
    "target_ast": Object {
      "value": 0.01486546751895347,
    },
  },
  "scrambleAssisted": Object {},
  "totalAssists": 147,
  "totalPlaysMade": 672.7,
  "transition": Object {
    "source_3p_ast": Object {
      "value": 0.034190575293592984,
    },
    "source_mid_ast": Object {
      "value": 0.002973093503790694,
    },
    "source_rim_ast": Object {
      "value": 0.04459640255686041,
    },
    "source_sf": Object {
      "value": 0.060019325107774635,
    },
    "source_to": Object {
      "value": 0.02973093503790694,
    },
    "target_ast": Object {
      "value": 0.05648877657202318,
    },
  },
  "transitionAssisted": Object {},
  "unassisted": Object {
    "source_3p_ast": Object {
      "value": 0.20662999851345323,
    },
    "source_mid_ast": Object {
      "value": 0.09662553887319755,
    },
    "source_rim_ast": Object {
      "value": 0.1486546751895347,
    },
    "source_sf": Object {
      "value": 0.14969525791586144,
    },
    "source_to": Object {
      "value": 0.10257172588077894,
    },
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerStyle 3`] = `
Object {
  "assisted": Object {
    "source_3p_ast": Object {
      "value": 35,
    },
    "source_mid_ast": null,
    "source_rim_ast": Object {
      "value": 17,
    },
    "target_ast": Object {
      "value": 147,
    },
  },
  "assistedMissAdjustments": undefined,
  "scramble": Object {
    "source_3p_ast": Object {
      "value": 6,
    },
    "source_mid_ast": Object {
      "value": 1,
    },
    "source_rim_ast": Object {
      "value": 4,
    },
    "source_sf": Object {
      "value": 5.699999999999999,
    },
    "target_ast": Object {
      "value": 10,
    },
  },
  "scrambleAssisted": Object {},
  "totalAssists": 1,
  "totalPlaysMade": 1,
  "transition": Object {
    "source_3p_ast": Object {
      "value": 8,
    },
    "source_mid_ast": Object {
      "value": 1,
    },
    "source_rim_ast": Object {
      "value": 17,
    },
    "source_sf": Object {
      "value": 40.375,
    },
    "target_ast": Object {
      "value": 38,
    },
  },
  "transitionAssisted": Object {},
  "unassisted": Object {
    "source_3p_ast": Object {
      "value": 21,
    },
    "source_mid_ast": Object {
      "value": 18,
    },
    "source_rim_ast": Object {
      "value": 48,
    },
    "source_sf": Object {
      "value": 100.69999999999999,
    },
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPlayerStyle 4`] = `
Object {
  "assisted": Object {
    "source_3p_ast": Object {
      "value": 0.15608740894901144,
    },
    "source_mid_ast": null,
    "source_rim_ast": Object {
      "value": 0.0505425895644418,
    },
    "target_ast": Object {
      "value": 0.218522372528616,
    },
  },
  "assistedMissAdjustments": undefined,
  "scramble": Object {
    "source_3p_ast": Object {
      "value": 0.026757841534116245,
    },
    "source_mid_ast": Object {
      "value": 0.002973093503790694,
    },
    "source_rim_ast": Object {
      "value": 0.011892374015162776,
    },
    "source_sf": Object {
      "value": 0.016352014270848817,
    },
    "source_to": Object {
      "value": 0.002973093503790694,
    },
    "target_ast": Object {
      "value": 0.01486546751895347,
    },
  },
  "scrambleAssisted": Object {},
  "totalAssists": 147,
  "totalPlaysMade": 672.7,
  "transition": Object {
    "source_3p_ast": Object {
      "value": 0.03567712204548833,
    },
    "source_mid_ast": Object {
      "value": 0.002973093503790694,
    },
    "source_rim_ast": Object {
      "value": 0.0505425895644418,
    },
    "source_sf": Object {
      "value": 0.1040582726326743,
    },
    "source_to": Object {
      "value": 0.02973093503790694,
    },
    "target_ast": Object {
      "value": 0.05648877657202318,
    },
  },
  "transitionAssisted": Object {},
  "unassisted": Object {
    "source_3p_ast": Object {
      "value": 0.09365244536940687,
    },
    "source_mid_ast": Object {
      "value": 0.05351568306823249,
    },
    "source_rim_ast": Object {
      "value": 0.1427084881819533,
    },
    "source_sf": Object {
      "value": 0.2556860413259997,
    },
    "source_to": Object {
      "value": 0.10257172588077894,
    },
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPosCategoryAssistNetwork 1`] = `
Array [
  Object {
    "info": Object {
      "order": 0,
    },
    "order": 0,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 1,
      "source_3p_ast": Object {
        "extraInfo": Array [
          "slash-and-kick",
          "hockey assist after defense collapses inside",
          "misc action",
        ],
        "value": 0.09050943884147919,
      },
      "source_rim_ast": Object {
        "extraInfo": Array [
          "cut",
        ],
        "value": 0.0439617274372899,
      },
      "target_3p_ast": Object {
        "extraInfo": Array [
          "drive-and-kick",
          "hockey assist after defense collapses inside",
          "misc action",
        ],
        "value": 0.4761904761904762,
      },
      "target_3p_efg": Object {
        "extraInfo": undefined,
        "value": 0.13601368691189053,
      },
      "target_ast": Object {
        "extraInfo": undefined,
        "value": 0.3801396431342126,
      },
      "target_mid_ast": Object {
        "extraInfo": Array [
          "spread offense",
          "misc action",
          "zone buster",
        ],
        "value": 0.06802721088435373,
      },
      "target_mid_efg": Object {
        "extraInfo": undefined,
        "value": 0.0972972972972973,
      },
      "target_rim_ast": Object {
        "extraInfo": Array [
          "cut",
        ],
        "value": 0.4557823129251701,
      },
      "target_rim_efg": Object {
        "extraInfo": undefined,
        "value": 0.07437389324563622,
      },
    },
    "order": 1,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 2,
    },
    "order": 2,
    "score": 0,
  },
]
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPosCategoryAssistNetwork 2`] = `
Array [
  Object {
    "info": Object {
      "order": 0,
    },
    "order": 0,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 1,
      "source_3p_ast": Object {
        "extraInfo": undefined,
        "value": 0.09050943884147919,
      },
      "source_rim_ast": Object {
        "extraInfo": undefined,
        "value": 0.0439617274372899,
      },
      "target_3p_ast": Object {
        "extraInfo": undefined,
        "value": 0.4761904761904762,
      },
      "target_3p_efg": Object {
        "extraInfo": undefined,
        "value": 0.13601368691189053,
      },
      "target_ast": Object {
        "extraInfo": undefined,
        "value": 0.3801396431342126,
      },
      "target_mid_ast": Object {
        "extraInfo": undefined,
        "value": 0.06802721088435373,
      },
      "target_mid_efg": Object {
        "extraInfo": undefined,
        "value": 0.0972972972972973,
      },
      "target_rim_ast": Object {
        "extraInfo": undefined,
        "value": 0.4557823129251701,
      },
      "target_rim_efg": Object {
        "extraInfo": undefined,
        "value": 0.07437389324563622,
      },
    },
    "order": 1,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 2,
    },
    "order": 2,
    "score": 0,
  },
]
`;

exports[`PlayTypeUtils PlayTypeUtils - buildPosCategoryAssistNetwork 3`] = `
Array [
  Object {
    "info": Object {
      "order": 0,
    },
    "order": 0,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 1,
      "source_3p_ast": Object {
        "extraInfo": Array [
          "drive-and-kick",
          "hockey assist after defense collapses inside",
          "misc action",
        ],
        "value": 0.09050943884147919,
      },
      "source_rim_ast": Object {
        "extraInfo": Array [
          "cut",
        ],
        "value": 0.0439617274372899,
      },
      "target_3p_ast": Object {
        "extraInfo": Array [
          "slash-and-kick",
          "hockey assist after defense collapses inside",
          "misc action",
        ],
        "value": 0.4761904761904762,
      },
      "target_3p_efg": Object {
        "extraInfo": undefined,
        "value": 0.13601368691189053,
      },
      "target_ast": Object {
        "extraInfo": undefined,
        "value": 0.3801396431342126,
      },
      "target_mid_ast": Object {
        "extraInfo": Array [
          "spread offense",
          "misc action",
        ],
        "value": 0.06802721088435373,
      },
      "target_mid_efg": Object {
        "extraInfo": undefined,
        "value": 0.0972972972972973,
      },
      "target_rim_ast": Object {
        "extraInfo": Array [
          "cut",
        ],
        "value": 0.4557823129251701,
      },
      "target_rim_efg": Object {
        "extraInfo": undefined,
        "value": 0.07437389324563622,
      },
    },
    "order": 1,
    "score": 0,
  },
  Object {
    "info": Object {
      "order": 2,
    },
    "order": 2,
    "score": 0,
  },
]
`;

exports[`PlayTypeUtils PlayTypeUtils - enrichUnassistedStats 1`] = `
Object {
  "source_3p_ast": Object {
    "extraInfo": Array [
      "dribble jumper off misc action",
      "off ball-screen",
      "dribble jumper off ISO",
    ],
    "value": 0.05430566330488751,
  },
  "source_mid_ast": Object {
    "extraInfo": Array [
      "drive and pull-up off ball-screen or ISO",
      "misc action",
    ],
    "value": 0.0465477114041893,
  },
  "source_rim_ast": Object {
    "extraInfo": Array [
      "attacks the rim off pick-and-roll",
      "ISO",
    ],
    "value": 0.12412723041117145,
  },
  "source_sf": Object {
    "extraInfo": Array [
      "fouled driving to the rim",
      "fouled in the bonus",
      "fouled cutting",
    ],
    "value": 0.2604085854667701,
  },
}
`;

exports[`PlayTypeUtils PlayTypeUtils - enrichUnassistedStats 2`] = `
Object {
  "source_3p_ast": Object {
    "extraInfo": Array [
      "dribble jumper off misc action",
      "off ball-screen",
      "dribble jumper off ISO",
    ],
    "value": 0.05430566330488751,
  },
  "source_mid_ast": Object {
    "extraInfo": Array [
      "drive and pull-up off ball-screen or ISO",
      "misc action",
    ],
    "value": 0.0465477114041893,
  },
  "source_rim_ast": Object {
    "extraInfo": Array [
      "attacks the rim off pick-and-roll",
      "ISO",
    ],
    "value": 0.12412723041117145,
  },
  "source_sf": Object {
    "extraInfo": Array [
      "fouled driving to the rim",
      "fouled in the bonus",
      "fouled cutting",
    ],
    "value": 0.2604085854667701,
  },
}
`;
